packages/node/src/indexer/project.service.spec.ts

```diff
@@ -60,79 +54,143 @@ const demoProjects = [
-      null as unknown as any,
-    );
+    const module = await Test.createTestingModule({
+      providers: [
+        ConnectionPoolStateManager,
+        ConnectionPoolService,
+        {
+          provide: 'ISubqueryProject',
+          useFactory: () => p,
+        },
+        {
+          provide: NodeConfig,
+          useFactory: () => ({}),
+        },
+        {
+          provide: ProjectService,
+          useFactory: (apiService: ApiService, project: SubqueryProject) =>
+            new ProjectService(
+              {
+                validateProjectCustomDatasources: jest.fn(),
+              } as unknown as DsProcessorService,
+              apiService,
+              null as unknown as any,
+              null as unknown as any,
+              { query: jest.fn() } as unknown as any,
+              project,
+              projectUpgrade,
+              {
+                initCoreTables: jest.fn(),
+                storeCache: {
+                  metadata: {},
+                  flushCache: jest.fn(),
+                  _flushCache: { bind: jest.fn() },
+                },
+              } as unknown as any,
+              { unsafe: false } as unknown as NodeConfig,
+              {
+                getDynamicDatasources: jest.fn().mockResolvedValue([]),
+                init: jest.fn(),
+              } as unknown as DynamicDsService,
+              null as unknown as any,
+              null as unknown as any,
+            ),
+          inject: [ApiService, 'ISubqueryProject'],
+        },
+        EventEmitter2,
+        ApiService,
+        {
+          provide: ProjectUpgradeService,
+          useValue: projectUpgrade,
+        },
+      ],
+      imports: [EventEmitterModule.forRoot()],
+    }).compile();
+
+    const app = module.createNestApplication();
+    await app.init();
+    apiService = app.get(ApiService);
+    await apiService.init();
+    projectUpgradeService = app.get(
+      ProjectUpgradeService,
+    ) as ProjectUpgradeService<SubqueryProject>;
 
+    (projectUpgradeService as any).updateIndexedDeployments = jest.fn();
+    (projectUpgradeService as any).getDeploymentsMetadata = jest
+      .fn()
+      .mockResolvedValue({ 1: 'project1', 4: 'project4' } as Record<
+        number,
+        string
+      >);
+    (projectUpgradeService as any).validateIndexedData = jest
+      .fn()
+      .mockReturnValue(5);
+    projectService = module.get(ProjectService);
     // Mock db related returns
-    (service as any).ensureProject = jest.fn().mockResolvedValue('mock-schema');
-    (service as any).ensureMetadata = jest.fn();
-    (service as any).initDbSchema = jest.fn();
-    (service as any).initUnfinalizedInternal = jest
+    (projectService as any).ensureProject = jest
+      .fn()
+      .mockResolvedValue('mock-schema');
+    (projectService as any).ensureMetadata = jest.fn();
+    (projectService as any).initDbSchema = jest.fn();
+    (projectService as any).initUnfinalizedInternal = jest
```

Comment:
Why not just create a Testing project service class that extends the BaseProjectService?